content/post/graphql-with-go-and-postgresql.md

```diff
@@ -0,0 +1,377 @@
+# Query to retrieve all data about all the posts of user1.
+curl -XPOST http://vm:8080/graphql -d '{user(id:1){posts{id,title,body}}}'
+# Query to retrive the user who wrote post2, if post2 was written by user1; a contrived example by displays the flexibility of the language.
+curl -XPOST http://vm:8080/graphql -d '{user(id:1){post(id:2){user{id,email}}}}'
+```
+_Listing 2_
+
+By carefully desiging the mutations and type attributes, powerfull and expressive queries are possible.
+
+## Designing the PostgreSQL database
+
+The relational database design is, as usual, driven by the need to avoid data duplication.This approach was chosen for two reasons:
+ 1. to show that there is no need for a specialized database technology or to learn and use new design techniques to accommodate a GraphQL API.
+ 2. to show that a GraphQL API can still be created on top of existing databases, more specifically databases designed to back REST endpoints or even traditional server-side rendered html websites.
+
+See _Appendix 1_ for a discussion on differences between relational and graph databases with respect to building a GraphQL API. _Listing 3_ shows the SQL commands to create the new database. The database schema generally matches the GraphQL schema. The `followers` relation needed to be added to support the `follow/unfollow` mutations.
+
+```sql
+CREATE TABLE IF NOT EXISTS users (
+  id SERIAL PRIMARY KEY,
+  email VARCHAR(100) NOT NULL
+);
+CREATE TABLE IF NOT EXISTS posts (
+  id SERIAL PRIMARY KEY,
+  user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+  title VARCHAR(200) NOT NULL,
+  body TEXT NOT NULL
+);
+CREATE TABLE IF NOT EXISTS comments (
+  id SERIAL PRIMARY KEY,
+  user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+  post_id INTEGER NOT NULL REFERENCES posts(id) ON DELETE CASCADE,
+  title VARCHAR(200) NOT NULL,
+  body TEXT NOT NULL
+);
+CREATE TABLE IF NOT EXISTS followers (
+  follower_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+  followee_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+  PRIMARY KEY(follower_id, followee_id)
+);
+```
+_Listing 3_
+
+## Golang API Implementation
+
+The graphql parser implemented in go and used in this project is `github.com/graphql-go/graphql`. It contains a query parser, but no schema parser. This requires the programer to build the GraphQL schema in go using the constructs offered by the library. This is unlike the reference [nodejs implementation](https://github.com/graphql/graphql-js), which offers a schema parser and exposes hooks for data fetching. As such the schema in `Listing 1` is only useful as a guideline and has to be translated into golang code. However, this _"limitation"_ offers the opportunity to peer behind the levels of abstraction and see how the schema relates to the graph traversal model for retrieving data. _Listing 4_ shows the implementation of the `Comment` vertex type:
```

Comment:
- **GraphQL** parser implemented in **Go**
- the **programmer**, in **Go**
- **Golang** code